src/wpilib/i2c.rs

```diff
@@ -0,0 +1,116 @@
+use wpilib::wpilib_hal::*;
+
+use std::slice;
+
+/// Helper enum that describes one of two possible locations on the roborio for i2c devices
+#[derive(Copy, Clone)]
+pub enum I2cPort {
+    /// Enum option for i2c device plugged into the onboard i2c connectors on the roborio
+    OnBoard = 0,
+    /// Enum option for i2c device, such as a navX board, connected through the MXP port
+    MXP = 1,
+}
+
+/// Describes various errored or otherwise special cases that may result from an i2c transaction
+pub enum I2cState {
+    // TODO find is there is any way to get more i2c transaction information out of HAl than just
+    // success or failure
+
+    /// Indicates a general error in an i2c transaction
+    IOError,
+    /// Indicates when no string, or an invalid string was received from the i2c transaction
+    InvalidReceiveString,
+    /// HAL i2c functions return -1 on abort
+    TransferAbort,
+    /// HAL i2c functions return >= 0 when successful
+    Success,
+}
+
+/// Struct for sending and receiving data over i2c
+pub struct I2cInterface {
+    port: I2cPort,
+    device_address: i32,
+}
+
+impl I2cInterface {
+    /// users create a new I2cInterface here
+    pub fn new(p: I2cPort, addr: i32, init_status: &mut i32) -> I2cInterface {
+        unsafe { HAL_InitializeI2C(p as i32, init_status); };
+        I2cInterface { port: p, device_address: addr }
+    }
+
+    /// base function for a single i2c transaction, in which both parties may send and receive data
+    /// note: unlike wpilib proper, transaction returns a new String as the result, and consumes the str parameter
+    pub fn transaction(&self, sent: &str, send_size: i32, received: &str, receive_size: i32) -> Result<String, I2cState> {
+        let send_string = String::from(sent);
```

Comment:
I'd really rather not use `String` here for several reasons:
- Rust's `String` type is UTF-8, which obviously does not map properly to raw bytes. This is _not_ what you want.
- `String` allocates memory on the heap, which is unacceptable for realtime applications and especially so for a library designed to be used in such applications. `malloc` (or `jemalloc`) can cause an unbounded delay, which can and will cause you to miss timing guarantees. This is the reasoning behind the use of byte arrays. I would suggest the following function signatures:
```rust
pub fn transaction(&mut self, send: &[u8], recv: &mut [u8]) -> Result<(), I2cError>;
pub fn read(&self, recv: &mut [u8]) -> Result<(), I2cError>;
pub fn write(&mut self, send: &[u8]) -> Result<(), I2cError>;
```
Note that I suggested a few changes besides the discarding of `String`: I would prefer the `send_size` to be encoded in the slice, which enforces validity at compile-time (i.e. you can't try to send more bytes than you have), and `received` should be a mutable reference to a byte array. The reason I would prefer this to a return value mostly has to do with the allocation I mentioned earlier - if you don't have the user allocate it themselves, then the function has to do a heap allocation. Also, `I2cState` is not a reasonable type to use as `Err`, as it contains `Success`. I would prefer it renamed to `I2cError`, so that `Result<(), I2cError>` could be used reasonably as the return type. Finally, `transaction` and `write` should be mutable functions (`&mut self`), as while they don't directly mutate the object's state, they do have side-effects on the physical object and effectively do mutate state.

With these changes, you would still be able to call the function like you were passing in a string, but as a byte array:
```rust
let received = [0; 9];
i2c.transaction(b"some data", &mut received)?;
```
This ended up being a little longer than I expected - sorry about that. However, I do feel strongly about `String` not being the proper datatype for use here. What scenario(s) prompt you to prefer it over a simple byte array?